hypershift-operator/controllers/hostedcluster/hostedcluster_webhook.go

```diff
@@ -1,22 +1,85 @@
 package hostedcluster
 
 import (
+	"context"
 	"fmt"
 
 	hyperv1 "github.com/openshift/hypershift/api/v1beta1"
+	"github.com/openshift/hypershift/support/supportedversion"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
 	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-// SetupWebhookWithManager sets up webhooks.
+type hostedClusterDefaulter struct {
+}
+
+type nodePoolDefaulter struct {
+	client client.Client
+}
+
+func (defaulter *hostedClusterDefaulter) Default(ctx context.Context, obj runtime.Object) error {
+	hcluster, ok := obj.(*hyperv1.HostedCluster)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected a HostedCluster but got a %T", obj))
+	}
+
+	if hcluster.Spec.Release.Image != "" {
+		return nil
+	}
+
+	pullSpec, err := supportedversion.LookupLatestSupportedRelease(ctx)
+	if err != nil {
+		return fmt.Errorf("unable to find default release image: %w", err)
+	}
+	hcluster.Spec.Release.Image = pullSpec
+
+	return nil
+}
+
+func (defaulter *nodePoolDefaulter) Default(ctx context.Context, obj runtime.Object) error {
+	np, ok := obj.(*hyperv1.NodePool)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected a NodePool but got a %T", obj))
+	}
+
+	if np.Spec.Release.Image != "" {
+		return nil
+	} else if np.Spec.ClusterName == "" {
```

Comment:
Why is the defaulter doing throwing this validation error? the .ClusterName is already required an immutable

Reply:
Two reasons

1. I don't think ClusterName required to be a non empty string. I believe it is only required to be immutable. this is the validation rule for cluster name `+kubebuilder:validation:XValidation:rule="self == oldSelf", message="ClusterName is immutable"`

2. The mutation webhook runs before the validation, which is why we can default the empty release image even though it is required to not be empty via this validation rule `+kubebuilder:validation:Pattern=^(\w+\S+)$`


So, i'm kind of in a weird spot with the cluster name here.

Reply:
you're right though this ensures presence but it would accept an explicitly set zero value for the string, and as you mentioned if we enforced != "" validation at the crd schema it would run after this, so I'm fine either way.